### PR TITLE
`karmadactl`: Add the reserved label `karmada.io/system` to resources created by the `join` command

### PR DESCRIPTION
**Pull request by a7i**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
Follow-up to https://github.com/karmada-io/karmada/pull/3262

Adds the well-known "karmada.io/managed" label to resources created by karmada controllers

We need a way to bypass creation of some ClusterRole/Rolebindings by our OPA Gatekeeper and we don't have a consistent label to select on. When we join a cluster via push command, it creates ClusterRole/ClusterRoleBinding that are too open (`*` and `*`) which is by default blocked by our opa-gatekeeper policies. This allows us to bypass it by the karmada managed label.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
`karmadactl`: Add the reserved label `karmada.io/system` to resources created by the `join` command.
```

### CODE REVIEW

**Review by grosser — APPROVED** [review]
✨

File: pkg/util/credential.go
Comment (zhzhuang-zju) [inline pkg/util/credential.go:53]: The variable `managedByKarmadaLabels` is extracted here. Would it be better if it was exported so that other files can also use it? Like, https://github.com/karmada-io/karmada/blob/f03b5e9803e4738706bb6fcf807350a55a4ad703/pkg/controllers/certificate/cert_rotation_controller.go#L239-L245 could be changed to 
```go
	certificateSigningRequest := &certificatesv1.CertificateSigningRequest{
		ObjectMeta: metav1.ObjectMeta{
			Name: csrName,
			Labels: util.ManagedByKarmadaLabels,
			},
		},
```

File: pkg/util/credential.go
Comment (chaunceyjiang) [inline pkg/util/credential.go:72]: https://github.com/karmada-io/karmada/blob/adef1e59748e1e1d31cb75fffe406b5dd69a66d7/pkg/util/serviceaccount.go#L104-L114
Does the `secret` here also need to be handled in the same way?

Comment (RainbowMango) [reply pkg/util/credential.go:72]: Nice finding! I think Yes.

Comment (chaunceyjiang) [reply pkg/util/credential.go:72]: /cc @a7i

Comment (XiShanYongYe-Chang) [reply pkg/util/credential.go:72]: Kindly ping @a7i

Comment (a7i) [reply pkg/util/credential.go:72]: nice catch, let me fix that and rebase

File: pkg/util/credential.go
Comment (a7i) [reply pkg/util/credential.go]: Good suggestion, implemented ✅

File: pkg/util/credential.go
Comment (a7i) [reply pkg/util/credential.go:67]: Implemented ✅

File: pkg/karmadactl/register/register.go
Comment (a7i) [inline pkg/karmadactl/register/register.go:381]: left this since it goes nicely with the comment above it

**Comment by XiShanYongYe-Chang** [comment]
Hi @chaunceyjiang, can you help take a review again?

**Review by chaunceyjiang — COMMENTED** [review]
Thanks~


/LGTM

**Review by XiShanYongYe-Chang — COMMENTED** [review]
/lgtm
/cc @RainbowMango

**Review by RainbowMango — COMMENTED** [review]
/assign
Another PR I missed :( :(   I'll look at it ASAP.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

Forced pushed to solve conflicts and updated release-notes.

File: pkg/controllers/certificate/cert_rotation_controller.go
Comment (whitewindmills) [inline pkg/controllers/certificate/cert_rotation_controller.go:242]: /hold
since it's ceated in the Karmada control-plane, there is no need to add this label here.

File: pkg/controllers/cluster/cluster_controller.go
Comment (whitewindmills) [inline pkg/controllers/cluster/cluster_controller.go:426]: ditto

**Review by whitewindmills — COMMENTED** [review]
isn’t this label `karmada.io/managed` used to add to resources in the member cluster? this indicates that the resources on the member cluster are managed by Karmada, so Karmada can update or delete them.
https://github.com/karmada-io/karmada/blob/b217a9ce2d6cc042725d370952696c31d03778ca/pkg/util/objectwatcher/objectwatcher.go#L303-L305

File: pkg/controllers/certificate/cert_rotation_controller.go
Comment (RainbowMango) [reply pkg/controllers/certificate/cert_rotation_controller.go:242]: Good point!
@a7i do you agree with that?

Comment (a7i) [reply pkg/controllers/certificate/cert_rotation_controller.go:242]: when I added this label, my intention was to exclude it from opa-gatekeepr policies. for our use-case, we still run opa at the api level (otherwise, the rollouts would pass, but propagation would fail).

**Comment by XiShanYongYe-Chang** [comment]
Kindly ping @a7i

**Comment by a7i** [comment]
Hi @XiShanYongYe-Chang I commented [here](https://github.com/karmada-io/karmada/pull/4620#discussion_r1603280104) on the rationale / use-case

**Comment by grosser** [comment]
@XiShanYongYe-Chang and @whitewindmills 
would love to get this moving again
... we need a label on these resources, it does not really matter which
afaik ManagedByKarmadaLabelValue would make sense, but if we can unblock with a new "ManagedByKarmadaControllerLabel" then that's also fine, just need a clear decision

**Comment by a7i** [comment]
happy to adjust the PR based on feedback from karmada team.

- do you suggest adding a new label? if so, what should key/value be?

**Comment by XiShanYongYe-Chang** [comment]
We need to express two semantics, one to indicate that resources in the member clusters are managed by the Karmada system, and the other to indicate that resources on the Karmada control plane are managed by the Karmada system. I propose to distinguish between these two cases, so I came up with two ways:
1. Use two labels to represent this, for example: `karmada.io/system: true` and `karmada.io/managed: true`
2. Use different label values to distinguish it, for example: `karmada.io/manager: member-cluster` and `karmada.io/manager: control-plane`

Ask more ideas and suggestions from @chaunceyjiang @RainbowMango @whitewindmills @zhzhuang-zju

**Comment by whitewindmills** [comment]
solution 1 looks good to me.

**Comment by RainbowMango** [comment]
I'll revisit this tomorrow and get back to you then.

> In addition, if we have another choice for OPA to bypass?

What I'm thinking is if we want a label we should make sure every resource that was created during the joining process should have it, not just those credentials resources.

**Comment by RainbowMango** [comment]
> 1 Use two labels to represent this, for example: karmada.io/system: true and karmada.io/managed: true
2 Use different label values to distinguish it, for example: karmada.io/manager: member-cluster and karmada.io/manager: control-plane

I vote for option 1, too.

Please @a7i update this PR if you also agree with it.

**Comment by XiShanYongYe-Chang** [comment]
Kindly ping @a7i 

I will also update the labels on the Endpointslice collected on the control plane in sync.

**Comment by a7i** [comment]
@XiShanYongYe-Chang no rush, but should be ready for your review

**Comment by XiShanYongYe-Chang** [comment]
OK, thanks~
/retest

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks~

File: pkg/util/constants.go
Comment (a7i) [reply pkg/util/constants.go:266]: Implemented per this comment from five months ago: https://github.com/karmada-io/karmada/pull/4620#discussion_r1493130385

Happy to revert to the way it was

Comment (XiShanYongYe-Chang) [reply pkg/util/constants.go:266]: I suggest using the way it was before the PR was submitted.

Comment (a7i) [reply pkg/util/constants.go:266]: so to be clear:

re: `karmadaAgentLabels` was a map before the PR was submitted so leave as-is: https://github.com/karmada-io/karmada/blob/71de3dc02247e22da249c285fa0c0c032ae01445/pkg/karmadactl/register/register.go#L115

re: `KarmadaManagedLabels` and disregard @zhzhuang-zju suggestion and create maps inline as opposed to reusing one?

re: `KarmadaSystemLabels` follow managed labels

Comment (zhzhuang-zju) [reply pkg/util/constants.go:266]: @a7i @XiShanYongYe-Chang It's great to have a discussion about how to make the code better~ In fact, this question can be divided into two parts: 1. whether it is necessary to wrap the label as a map. 2. if it is wrapped as a map, whether it is necessary to export it. refer to https://github.com/karmada-io/karmada/pull/4620#discussion, I meant that if it is wrapped as a map, it is possible to export it for better scalability. I think what @XiShanYongYe-Chang means is that step 1 is not necessary because there is only one label, so there is no need to wrap it into a map. That's something I agree with, and I think the final opinion is that we can leave merged code alone, and the new code does not need to wrap the label into a map again. What do you think?

Comment (a7i) [reply pkg/util/constants.go:266]: Can you please be more specific with the three labels to avoid back and forth?

Comment (XiShanYongYe-Chang) [reply pkg/util/constants.go:266]: Just make it keep with before will be ok, just like this:
```go
var karmadaAgentLabels = map[string]string{"app": KarmadaAgentName}

func XXX(){
       ...
        ObjectMeta: metav1.ObjectMeta{
	        Name:      KarmadaAgentName,
	        Namespace: o.Namespace,
        	Labels:    karmadaAgentLabels,
        },
       ...
}
```

Comment (zhzhuang-zju) [reply pkg/util/constants.go:266]: combine with https://github.com/karmada-io/karmada/pull/4620#discussion_r1673916240 and https://github.com/karmada-io/karmada/pull/4620#discussion_r1673921429 is ok. @a7i ,If we didn't make it clear, please let us know

Comment (a7i) [reply pkg/util/constants.go:266]: >just replaced with labelName: labelvalue wherever used.

It's not very DRY, but at this point it's been 5 months and I'd like to get the PR moving. Commit coming up soon

Comment (a7i) [reply pkg/util/constants.go]: I hope I got it right this time 🤞🏼 

ready for your review

**Comment by XiShanYongYe-Chang** [comment]
Thanks~
/lgtm

**Comment by zhzhuang-zju** [comment]
/lgtm
and failed e2e cases also rerun successfully~
cc @whitewindmills I observed that you added the `/hold` label earlier, please take a look at it again for any other comments? Note that once `hold` is canceled, pr will get merged since there is `approve` label

File: pkg/util/constants.go
Comment (whitewindmills) [inline pkg/util/constants.go:55]: we'd better keep the first letter capitalized for `Karmada` in comments. others look good to me.
@RainbowMango do you wanna review it again?

Comment (RainbowMango) [reply pkg/util/constants.go:55]: no big deal.
I'm looking at it.

**Pull request merged**
